anvio/threadingops.py

```diff
@@ -0,0 +1,551 @@
+            if isinstance(thread.target_return_value, CommandError):
+                if self.logger:
+                    self.logger.progress.end()
+                raise thread.target_return_value
+
+    @staticmethod
+    def _remove_files(paths: Collection[str]) -> None:
+        """Helper function to remove all the given `paths`."""
+        for path in paths:
+            if filesnpaths.is_file_exists(path, dont_raise=True):
+                os.remove(path)
+
+
+class ThreadedProdigalRunner(ThreadedCommandRunner):
+    def __init__(self,
+                 input_file_path: str,
+                 collated_output_file_paths: Mapping[str, str],
+                 number_of_splits: int,
+                 log_file_path: str,
+                 logger: Optional[Logger],
+                 installed_version: str,
+                 parser: Callable[[str], Dict[str, Any]],
+                 # while it will take str or int, it will be saved as a str
+                 translation_table: Optional[Union[str, int]]
+                 ) -> None:
+        """See ThreadedCommandRunner for more info.
+
+        Note:  Lots of state to set up here...use kwargs please!
+
+        Parameters not shared with Superclass
+        =====================================
+
+        - `installed_version`: The version string of Prodigal
+        - `parser`:  The Callable used to parse Prodigal peptide file deflines.  (See signature for expected types.)
+        - `translation_table`: Either a string or int specifying the specific translation table, or None.  If None, then
+           Prodigal's `-p meta` option will be used instead.
+        """
+        super().__init__(input_file_path=input_file_path,
+                         collated_output_file_paths=collated_output_file_paths,
+                         number_of_splits=number_of_splits,
+                         log_file_path=log_file_path,
+                         logger=logger)
+
+        self.installed_version = installed_version
+        self.parser = parser
+
+        # if it's an int we need to cast to str
+        if isinstance(translation_table, int):
+            translation_table = str(translation_table)
+
+        self.translation_table: Optional[str] = translation_table
+
+    # Implement the abstract methods
+    #
+    #
+    def _split_input_file(self) -> State:
+        """Split input fasta into the correct number of splits.
+
+        Returns `State` with the paths to each of the fasta splits.
+
+        Raises `ValueError` if `self.number_of_splits < 0.
+
+        See Superclass `_split_input_file` for more info.
+        """
+        # Todo: probably should move this check into the constructor.
+        if self.number_of_splits <= 0:
+            ValueError(f'number_of_splits muts be > 0.  Got {self.number_of_splits}')
+
+        # Todo are there errors to catch here?
+        self.input_file_splits = utils.split_fasta(self.input_file_path, parts=self.number_of_splits)
```

Comment:
I think `utils.split_fasta` needs a `randomize=False` flag and when it is set to `True` it should split the input FASTA file in such a way, resulting split files do not have the sequences in the same order.

What I mean is this: Currently when we split a FASTA file into two splits, the first split contains the first half of contigs, and the second split contains the second half. Assemblers often report contigs in an ordered fashion, where longest contigs appear at the beginning of the file. So if you are splitting an assembly output into 5 files, the first file may contain more nucleotides than the last four combined. Which kinda makes the multithreading a little less effective.

I realized this while testing this code on the infant gut dataset. I gave 6 threads, and 5 of them finished very quickly and there was only a single thread for more than 70% of the total runtime.

Reply:
Check out https://github.com/merenlab/anvio/pull/1448 for one solution to this.